File: sdk/src/agent.ts
import type { Signer } from "./types.js";
import { AbaPayError } from "./types.js";
import { DEFAULT_BASE_URL } from "./x402.js";

// ⚡ THE OTHER WAY IN — a reusable version of examples/agent-quickstart.mjs (verified live
// against production 2026-09-11: a throwaway key ran this exact request shape, got a real
// api_key back, called check_balance with it, then unlinked itself). Unlike x402.ts, this path
// links a wallet ONCE (a signature, not a deposit) and then calls the fuller MCP tool catalog
// with an api_key + a PIN on every payment — see https://agents.abapays.com/#a2a for why the
// two paths have genuinely different trust models.

export type ApprovedChain = "CELO" | "BASE";

export interface LinkParams {
  signer: Signer & { signMessage: (args: { message: string }) => Promise<`0x${string}`> };
  /** 4-6 digits. Yours to pick; AbaPay never generates or stores it in recoverable form. */
  pin: string;
  approvedChain?: ApprovedChain;
  approvedToken?: string;
  /** Shown in AbaPay's Agent Hub UI so a human can tell which integration minted this key. */
  label?: string;
  baseUrl?: string;
}

interface LinkResponse {
  success: boolean;
  api_key?: string;
  message?: string;
}

interface McpCallResponse {
  result?: { content?: { type: string; text?: string }[] };
  error?: { message?: string };
}

/**
 * A linked AbaPay agent session — one api_key, reusable across as many tool calls as needed.
 * Get one via {@link AbaPayAgent.link}.
 */
export class AbaPayAgent {
  private constructor(
    public readonly apiKey: string,
    public readonly walletAddress: string,
    private readonly baseUrl: string,
  ) {}

  /**
   * Prove ownership of a wallet with a plain signed message (no OAuth, no browser) and mint an
   * Agent Hub api_key. This does NOT grant AbaPay any spending allowance by itself — that's a
   * separate on-chain step (`approve` + `setSpendingAllowance`, see README.md's AbaPayV4
   * section or `examples/agent-quickstart.mjs`), deliberately left to the caller rather than
   * done here, since it moves the caller's own funds and shouldn't happen implicitly inside a
   * "just log in" call.
   *
   * @throws {AbaPayError} if the signature is rejected or the PIN is malformed.
   */
  static async link(params: LinkParams): Promise<AbaPayAgent> {
    const { signer, pin, approvedChain = "CELO", approvedToken, label, baseUrl = DEFAULT_BASE_URL } = params;
    if (!/^\d{4,6}$/.test(pin)) {
      throw new AbaPayError("PIN must be 4-6 digits.");
    }

    const timestamp = String(Date.now());
    const message = `AbaPay Agent Action: POST:/api/agent/link: ${timestamp}`;
    const signature = await signer.signMessage({ message });

    const res = await fetch(`${baseUrl.replace(/\/$/, "")}/api/agent/link`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-wallet-address": signer.address,
        "x-wallet-signature": signature,
        "x-wallet-timestamp": timestamp,
      },
      body: JSON.stringify({
        wallet_address: signer.address,
        channel: "MCP",
        pin,
        approved_chain: approvedChain,
        approved_token: approvedToken,
        mcp_key_label: label || "abapay-sdk",
      }),
    });
    const data = (await res.json()) as LinkResponse;
    if (!data.success || !data.api_key) {
      throw new AbaPayError(data.message || `Link failed (HTTP ${res.status}).`, undefined, data);
    }
    return new AbaPayAgent(data.api_key, signer.address, baseUrl);
  }

  /** Reattach to an api_key minted earlier (e.g. one saved from a previous {@link link} call) — no new signature needed. */
  static fromApiKey(apiKey: string, walletAddress: string, baseUrl = DEFAULT_BASE_URL): AbaPayAgent {
    return new AbaPayAgent(apiKey, walletAddress, baseUrl);
  }

  /** Low-level: call any MCP tool by name. `checkBalance`/`payBill`/etc. below cover the common ones with typed args. */
  async callTool(name: string, args: Record<string, unknown> = {}): Promise<string> {
    const res = await fetch(`${this.baseUrl.replace(/\/$/, "")}/api/mcp`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 1,
        method: "tools/call",
        params: { name, arguments: { api_key: this.apiKey, ...args } },
      }),
    });
    const data = (await res.json()) as McpCallResponse;
    if (data.error) throw new AbaPayError(data.error.message || "MCP tool call failed.", undefined, data);
    const text = data.result?.content?.find((c) => c.type === "text")?.text;
    if (text === undefined) throw new AbaPayError("MCP tool returned no text content.", undefined, data);
    return text;
  }

  checkBalance(chain: ApprovedChain = "CELO"): Promise<string> {
    return this.callTool("check_balance", { chain });
  }

  payBill(args: {
    pin: string;
    service: string;
    provider: string;
    account_number: string;
    amount_ngn: number;
    chain?: ApprovedChain;
    token?: string;
    variation_code?: string;
  }): Promise<string> {
    return this.callTool("pay_bill", args);
  }

  scheduleBill(args: Record<string, unknown>): Promise<string> {
    return this.callTool("schedule_bill", args);
  }

  transactionHistory(args: Record<string, unknown> = {}): Promise<string> {
    return this.callTool("transaction_history", args);
  }
}

File: python-sdk/src/abapay/agent.py
"""AbaPayAgent — mirrors sdk/src/agent.ts. Links a wallet ONCE (a signature, not a deposit)
and then calls the fuller MCP tool catalog with an api_key + a PIN on every payment. See
https://agents.abapays.com/a2a for why this and x402.py have genuinely different trust models.
"""

from __future__ import annotations

import json
import re
import time
import urllib.error
import urllib.request
from typing import Any, Optional

from .types import AbaPayError, LinkParams
from .x402 import DEFAULT_BASE_URL

_PIN_RE = re.compile(r"^\d{4,6}$")


def _http_json(url: str, method: str, body: Optional[dict] = None, headers: Optional[dict] = None) -> dict:
    data = json.dumps(body).encode("utf-8") if body is not None else None
    all_headers = {"Content-Type": "application/json"}
    if headers:
        all_headers.update(headers)
    req = urllib.request.Request(url, data=data, headers=all_headers, method=method)
    try:
        with urllib.request.urlopen(req) as res:
            return json.loads(res.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8")
        try:
            return json.loads(raw)
        except json.JSONDecodeError:
            raise AbaPayError(f"AbaPay returned a non-JSON response (HTTP {e.code}): {raw[:300]}")


class AbaPayAgent:
    """A linked AbaPay agent session — one api_key, reusable across as many tool calls as
    needed. Get one via `AbaPayAgent.link(...)`."""

    def __init__(self, api_key: str, wallet_address: str, base_url: str = DEFAULT_BASE_URL):
        self.api_key = api_key
        self.wallet_address = wallet_address
        self._base_url = base_url

    @staticmethod
    def link(params: LinkParams) -> "AbaPayAgent":
        """Prove ownership of a wallet with a plain signed message (no OAuth, no browser) and
        mint an Agent Hub api_key. This does NOT grant AbaPay any spending allowance by
        itself — that's a separate on-chain step (approve + setSpendingAllowance), left to the
        caller since it moves the caller's own funds.

        Raises AbaPayError if the signature is rejected or the PIN is malformed.
        """
        if not _PIN_RE.match(params.pin):
            raise AbaPayError("PIN must be 4-6 digits.")

        timestamp = str(int(time.time() * 1000))
        message = f"AbaPay Agent Action: POST:/api/agent/link: {timestamp}"
        signed = params.signer.sign_message(_personal_sign_message(message))
        signature = signed.signature.hex()
        if not signature.startswith("0x"):
            signature = "0x" + signature

        data = _http_json(
            f"{params.base_url.rstrip('/')}/api/agent/link",
            "POST",
            body={
                "wallet_address": params.signer.address,
                "channel": "MCP",
                "pin": params.pin,
                "approved_chain": params.approved_chain,
                "approved_token": params.approved_token,
                "mcp_key_label": params.label,
            },
            headers={
                "x-wallet-address": params.signer.address,
                "x-wallet-signature": signature,
                "x-wallet-timestamp": timestamp,
            },
        )
        if not data.get("success") or not data.get("api_key"):
            raise AbaPayError(data.get("message") or "Link failed.", response=data)
        return AbaPayAgent(data["api_key"], params.signer.address, params.base_url)

    @staticmethod
    def from_api_key(api_key: str, wallet_address: str, base_url: str = DEFAULT_BASE_URL) -> "AbaPayAgent":
        """Reattach to an api_key minted earlier — no new signature needed."""
        return AbaPayAgent(api_key, wallet_address, base_url)

    def call_tool(self, name: str, args: Optional[dict] = None) -> str:
        """Low-level: call any MCP tool by name. check_balance/pay_bill/etc below cover the
        common ones with typed args."""
        body = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "tools/call",
            "params": {"name": name, "arguments": {"api_key": self.api_key, **(args or {})}},
        }
        data = _http_json(f"{self._base_url.rstrip('/')}/api/mcp", "POST", body=body)
        if data.get("error"):
            raise AbaPayError(data["error"].get("message") or "MCP tool call failed.", response=data)
        content = (data.get("result") or {}).get("content") or []
        text = next((c.get("text") for c in content if c.get("type") == "text"), None)
        if text is None:
            raise AbaPayError("MCP tool returned no text content.", response=data)
        return text

    def check_balance(self, chain: str = "CELO") -> str:
        return self.call_tool("check_balance", {"chain": chain})

    def pay_bill(
        self,
        pin: str,
        service: str,
        provider: str,
        account_number: str,
        amount_ngn: float,
        chain: Optional[str] = None,
        token: Optional[str] = None,
        variation_code: Optional[str] = None,
    ) -> str:
        args: dict[str, Any] = {
            "pin": pin,
            "service": service,
            "provider": provider,
            "account_number": account_number,
            "amount_ngn": amount_ngn,
        }
        if chain:
            args["chain"] = chain
        if token:
            args["token"] = token
        if variation_code:
            args["variation_code"] = variation_code
        return self.call_tool("pay_bill", args)

    def schedule_bill(self, **args: Any) -> str:
        return self.call_tool("schedule_bill", args)

    def transaction_history(self, limit: Optional[int] = None, offset: Optional[int] = None) -> str:
        args: dict[str, Any] = {}
        if limit is not None:
            args["limit"] = limit
        if offset is not None:
            args["offset"] = offset
        return self.call_tool("transaction_history", args)


def _personal_sign_message(message: str):
    from eth_account.messages import encode_defunct

    return encode_defunct(text=message)
